File: src/main/java/lmller/github/io/gofkotlin/visitor/Visitor.java
package lmller.github.io.gofkotlin.visitor;

import java.util.Arrays;
import java.util.List;

public class Visitor {
    interface Shape {
        <T> T accept(ShapeVisitor<T> visitor);
    }

    public static class Square implements Shape {
        final double side;

        public Square(double side) {
            this.side = side;
        }

        @Override
        public <T> T accept(ShapeVisitor<T> visitor) {
            return visitor.visit(this);
        }
    }

    public static class Circle implements Shape {
        final double radius;

        public Circle(double radius) {
            this.radius = radius;
        }

        @Override
        public <T> T accept(ShapeVisitor<T> visitor) {
            return visitor.visit(this);
        }
    }

    public static class Rectangle implements Shape {
        final double width;
        final double height;

        public Rectangle(double width, double height) {
            this.width = width;
            this.height = height;
        }

        @Override
        public <T> T accept(ShapeVisitor<T> visitor) {
            return visitor.visit(this);
        }
    }

    interface ShapeVisitor<T> {
        T visit(Square element);

        T visit(Circle element);

        T visit(Rectangle element);
    }

    public static class AreaVisitor implements ShapeVisitor<Double> {
        public Double visit(Square element) {
            return element.side * element.side;
        }

        public Double visit(Circle element) {
            return Math.PI * element.radius * element.radius;
        }

        public Double visit(Rectangle element) {
            return element.height * element.width;
        }
    }

    public static class PerimeterVisitor implements ShapeVisitor<Double> {
        public Double visit(Square element) {
            return 4 * element.side;
        }

        public Double visit(Circle element) {
            return 2 * Math.PI * element.radius;
        }

        public Double visit(Rectangle element) {
            return (2 * element.height + 2 * element.width);
        }
    }

    public static void main(String[] args) {
        List<Shape> figures = Arrays.asList(new Circle(4), new Square(5), new Rectangle(6, 7));

        double totalArea = 0.0;
        ShapeVisitor<Double> areaVisitor = new AreaVisitor();
        for (Shape figure : figures) {
            totalArea += figure.accept(areaVisitor);
        }
        System.out.println("Total area = " + totalArea);

        double totalPerimeter = 0.0;
        ShapeVisitor<Double> perimeterVisitor = new PerimeterVisitor();
        for (Shape figure : figures) {
            totalPerimeter += figure.accept(perimeterVisitor);
        }
        System.out.println("Total perimeter = " + totalPerimeter);
    }
}

File: src/main/java/lmller/github/io/gofkotlin/visitor/Visitor.kt
package lmller.github.io.gofkotlin.visitor

sealed class Shape()
class Square(val side: Double) : Shape()
class Circle(val radius: Double) : Shape()
class Rectangle(val width: Double, val height: Double) : Shape()


fun main(args: Array<String>) {
    val figures = listOf(Circle(4.0), Square(5.0), Rectangle(6.0, 7.0))

    val perimeterVisitor = { shape: Shape ->
        when (shape) {
            is Rectangle -> 2 * shape.height + 2 * shape.width
            is Circle -> 2 * Math.PI * shape.radius
            is Square -> 4 * shape.side
        }
    }

    val areaVisitor = { shape: Shape ->
        when (shape) {
            is Rectangle -> shape.height * shape.width
            is Circle -> shape.radius.square() * Math.PI
            is Square -> shape.side.square()
        }
    }

    val totalArea = figures.sumByDouble { areaVisitor(it) }
    println("Total area = $totalArea")

    val totalPerimeter = figures.sumByDouble { perimeterVisitor(it) }
    println("Total perimeter = $totalPerimeter")
}

fun Double.square() = this * this